### PR TITLE
JIT exporter metrics

### PR DESCRIPTION
**Pull request by jack-berg**
[ExporterMetrics](https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/common/src/main/java/io/opentelemetry/exporter/internal/ExporterMetrics.java) standardizes metric collection from exporters. Instrumenting exporters presents a chicken and the egg problem since exporters are part of the SDK, yet rely on the SDK for instrumentation.

The effort to split out exporter configuration into SPI implementations in #4949 makes this more clear as exporter configuration code (like [jaeger configuration](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/SpanExporterConfiguration.java#L194)) sets the meter provider in a way that SPI implementations can't because they don't have access to the [meter provider](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/traces/ConfigurableSpanExporterProvider.java#L23). 

IMO instrumenting exporters is one of the few cases where depending on `GlobalOpenTelemetry` makes sense: Rather than worry about the ordering of initializing `SdkMeterProvider` and passing it to other exporters, just have those exporters instrument themselves using `GlobalOpenTelemetry#getMeterProvider()`. The one caveat is that instrumentation has to be "just in time", ensuring that `GlobalOpenTelemetry#set` has been called before instrumentation begins.

This PR adjusts `ExporterMetrics` instrumentation to be "just in time". It also adjusts all the exporters to use `GlobalOpenTelemetry.getMeterProvider` by default instead of `MeterProvider.noop()`.

### CODE REVIEW

**Review by trask — COMMENTED** [review]
what will happen if a user has configured `OpenTelemetrySdk` without registering it global?

**Comment by jack-berg** [comment]
> what will happen if a user has configured OpenTelemetrySdk without registering it global?

They have to manually call `setMeterProvider` on their respective exporters.

**Comment by trask** [comment]
> They have to manually call `setMeterProvider` on their respective exporters.

I guess, I'm wondering what happens if they don't do this, will the GlobalOpenTelemetry usage force another sdk to be auto-initialized?

**Comment by jack-berg** [comment]
> if they don't do this, will the GlobalOpenTelemetry usage force another sdk to be auto-initialized?

Ugh.. yes. Geez I'm really not a fan of [GlobalOpenTelemetry#maybeAutoconfigureAndSetGlobal](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java#L213-L240).

**Comment by mateuszrzeszutek** [comment]
> * Remove the `AutoConfiguredOpenTelemetrySdk` initialization side affect from `GlobalOpenTelemetry`. This seems controversial, but I'm not exactly sure why because I'm not sure who / what instrumentation relies on that side affect.

I prefer this option. We're discouraging everybody from using the `GlobalOpenTelemetry` in manually instrumented applications anyway, so why do we want it to be "smart" and auto-load the SDK? You don't need the global instance auto-configuring the SDK when using the agent, and when you're instrumenting the app manually you shouldn't be using the global in the first place.

**Comment by jack-berg** [comment]
With #5010 merged we don't need to worry about side affects of calling `GlobalOpenTelemetry.get()`. Autoconfiguration users will only be able to get instrumentation about their exporters if they use `AutoConfiguredOpenTelemetrySdk.builder().setResultAsGlobal(true)` or otherwise call `GlobalOpenTelemetry.set`. If not, exporters will use `OpenTelemetry.noop()`. #5021 unlocks exporter instrumentation in scenarios where `GlobalOpenTelemetry` is not set, but using `GlobalOpenTelemetry` seems find for now given the experimental state of this telemetry. 

@open-telemetry/java-approvers can you take another look?

**Review by trask — APPROVED** [review]

File: exporters/otlp/all/src/main/java/io/opentelemetry/exporter/otlp/metrics/OtlpGrpcMetricExporterBuilder.java [hunk 1]
```diff
@@ -57,6 +58,7 @@ public final class OtlpGrpcMetricExporterBuilder {
             DEFAULT_ENDPOINT,
             () -> MarshalerMetricsServiceGrpc::newFutureStub,
             GRPC_ENDPOINT_PATH);
+    delegate.setMeterProvider(MeterProvider.noop());
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/otlp/all/src/main/java/io/opentelemetry/exporter/otlp/metrics/OtlpGrpcMetricExporterBuilder.java:61]: This is actually needed to keep the behavior of `OtlpGrpcMetricExporterBuilder` the same as it is today. `delegate` is `GrpcExporterBuidler`, and has [changed](https://github.com/open-telemetry/opentelemetry-java/pull/4993/files#diff-8e59bf338a09156f05d1921b59b3be4e06d96307a98d60997cb45cfe3dc2bdeaR51) its behavior from using `MeterProvider.noop()` to `GlobalOpenTelemetry.getMeterProvider()`.

We currently don't record self-diagnostic metrics on metric exporters. I don't remember the exact reasoning, but I suspect its something about if metrics aren't exporting correctly then the diagnostic metrics won't export either so why collect them at all.

Same applies to `OtlpHttpGrpcExporterBuilder`.

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/ExporterMetrics.java [hunk 2]
```diff
@@ -25,45 +27,77 @@ public class ExporterMetrics {
...
+  }
+
+  private LongCounter exported() {
+    LongCounter exported = this.exported;
+    if (exported == null) {
+      exported = meter().counterBuilder(exporterName + ".exporter.exported").build();
+      this.exported = exported;
+    }
+    return exported;
+  }
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/common/src/main/java/io/opentelemetry/exporter/internal/ExporterMetrics.java:87]: It doesn't need to be synchronized. It looks a bit weird, but if two threads call `exported()` at the same time, both will initialize a counter and set `this.exported`, but this isn't a problem.

**Review by mateuszrzeszutek — APPROVED** [review]

**Comment by jack-berg** [comment]
@jkwatson any thoughts on this?

**Comment by jkwatson** [comment]
> @jkwatson any thoughts on this?

Ship it

**Review by jkwatson — APPROVED** [review]

**Comment by zeitlinger** [comment]
> We're discouraging everybody from using the `GlobalOpenTelemetry` in manually instrumented applications anyway

What is the recommended approach?

**Comment by jkwatson** [comment]
> > We're discouraging everybody from using the `GlobalOpenTelemetry` in manually instrumented applications anyway
> 
> What is the recommended approach?

Inject an instance of OpenTelemetry into the places where it's needed.

**Comment by zeitlinger** [comment]
> Inject an instance of OpenTelemetry into the places where it's needed.

Let me add some context: If I'm writing a library such as [riptide](https://github.com/zalando/riptide/blob/main/riptide-opentelemetry/src/main/java/org/zalando/riptide/opentelemetry/OpenTelemetryPlugin.java#L45) - what should I do?

**Pull request merged**